Initial commit for discussion bot

MengnanLi91 opened the pull request:
refer to #29986

This project originates from @hugary1995 's idea and his side project of creating a MOOSE Discussion database and retrieval method. Our goal is to integrate this into MOOSE and develop an automated MOOSE chatbot that displays the most relevant posts when a new discussion is initiated on MOOSE.

The current design utilizes purely vector retrieval(use LlamaIndex to build it). The algorithm compares the discussion title with the content of all previous discussions (title + discussions) in our database, and provides the most relevant posts to the user. In future updates, we plan to incorporate Large Language Models (LLMs) into this process to generate summaries of the recommended posts. The database will be updated regularly to include all new posts (potentially on a monthly basis). The scripts for fetching and generating the database, as well as the GitHub bot script, are included in this pull request. Unit tests for each script are also provided.

We are also looking for a fitting name for the discussion bot. If you have any suggestions, please share them here!

Below is a snapshot of how it looks like in discussion page:

<img width="1021" alt="Screenshot 2025-03-18 at 9 40 54 AM" src="https://github.com/user-attachments/assets/9f34cf93-7f05-4376-b01b-87f532536b06" />

link: https://github.com/MengnanLi91/moose/discussions/18

MengnanLi91 committed:
Initial commit for discussion bot #29986

Remove local path

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2720707/), step Size check on 1d1104e wanted to post the following:

Warning: This PR changes repo size by 14.23 MiB.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2720738/), step Docs: sync website on 1d1104e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29988/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2720754/), step Generate coverage on 1d1104e wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29988/coverage/xfem)

</details>



This comment will be updated on new commits.

MengnanLi91 commented:
@GiudGiud How do you want to set the threshold? If we don't find similar posts above this threshold, what should we reply to user?

GiudGiud commented:
> How do you want to set the threshold?

Experimentally? Start too low and raise it over time

> If we don't find similar posts above this threshold, what should we reply to user?

nothing. They can wait for someone to reply

GiudGiud commented on python/MooseDiscussion/db/encoding.npy:
can this be external? I dont like that we will be refreshing a 15MB file every now and then

MengnanLi91 replied:
Yes, I am discussing with @milljm and @loganharbour to move database to another repo. Also I am going to use llamaIndex and vector database rather than .npy and .csv. That will make the database smaller too.

milljm replied:
I wonder if there is a desire for a central storage for AI models in the Idaholab namespace. Just something to consider when we name this repo...

MengnanLi91 replied:
This is not AI or LLM, I just use open source vector database tool. I think we are good there.

MengnanLi91 committed:
Use llamaIndex for embedding and retrievial

use GITHUB_TOKEN

Need to be my name before merge

Add requirement file for conda enviroment

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2726290/), step Size check on 40b763f wanted to post the following:

Warning: This PR changes repo size by 25.83 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2726414/), step Size check on defbbea wanted to post the following:

Warning: This PR changes repo size by 25.83 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2737911/), step Size check on 0dc445e wanted to post the following:

Warning: This PR changes repo size by 25.83 MiB.

MengnanLi91 committed:
Codes for fetch, build database and recommend

Co-authored-by: Gary Hu <31555204+hugary1995@users.noreply.github.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2748785/), step Size check on a32f38d wanted to post the following:

Warning: This PR changes repo size by 25.84 MiB.

MengnanLi91 committed:
Update codes and database

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2749339/), step Size check on 4426c7c wanted to post the following:

Warning: This PR changes repo size by 13.48 MiB.

MengnanLi91 committed:
Update workflow to test on my repo

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2749390/), step Size check on 69c41b8 wanted to post the following:

Warning: This PR changes repo size by 13.48 MiB.

MengnanLi91 committed:
Update workflow to use enviroment variables

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2749441/), step Size check on 42ce72b wanted to post the following:

Warning: This PR changes repo size by 13.48 MiB.

MengnanLi91 committed:
Update GitHubbot dryrun and workflow

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2749503/), step Size check on be64170 wanted to post the following:

Warning: This PR changes repo size by 13.48 MiB.

MengnanLi91 committed:
Add enviromental variables to workflow

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2749554/), step Size check on 31b05d2 wanted to post the following:

Warning: This PR changes repo size by 13.48 MiB.

MengnanLi91 committed:
Update workflow to include enviromental variables

MengnanLi91 committed:
Put args in enviroment variables

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2751637/), step Size check on 7f68c72 wanted to post the following:

Warning: This PR changes repo size by 13.48 MiB.

MengnanLi91 committed:
Add embed model settings

MengnanLi91 committed:
use self.embed_model instead of embed_model

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2751740/), step Size check on e751c22 wanted to post the following:

Warning: This PR changes repo size by 13.48 MiB.

MengnanLi91 committed:
Set embeding model inside funciton

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2751792/), step Size check on 117d74f wanted to post the following:

Warning: This PR changes repo size by 13.48 MiB.

MengnanLi91 committed:
Make sure embedding model is set before call

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2751853/), step Size check on 0ed4fa5 wanted to post the following:

Warning: This PR changes repo size by 13.48 MiB.

MengnanLi91 committed:
Correct error message

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2751904/), step Size check on 8b3607d wanted to post the following:

Warning: This PR changes repo size by 13.48 MiB.

MengnanLi91 committed:
Correct model name

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2751955/), step Size check on 19dcd01 wanted to post the following:

Warning: This PR changes repo size by 13.48 MiB.

MengnanLi91 committed:
To be consistent with TOKEN name

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2752057/), step Size check on f45c917 wanted to post the following:

Warning: This PR changes repo size by 13.49 MiB.

MengnanLi91 committed:
Update the token name in workflow file

MengnanLi91 committed:
Formatting the generated response

MengnanLi91 committed:
Better formating, make url clickable

MengnanLi91 committed:
Add unit test for IndexGenerator.py

MengnanLi91 committed:
Doesn't need .env anymore

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2753585/), step Size check on baf50ce wanted to post the following:

Warning: This PR changes repo size by 13.49 MiB.

MengnanLi91 committed:
Update IndexGenerator unit tests

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2753777/), step Size check on d449f03 wanted to post the following:

Warning: This PR changes repo size by 13.49 MiB.

MengnanLi91 committed:
Update GitHubBot unit tests

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2754550/), step Size check on dc61268 wanted to post the following:

Warning: This PR changes repo size by 13.49 MiB.

MengnanLi91 committed:
Fixed two failed unit tests, one left

MengnanLi91 committed:
Update unit tests, add comments and dummy vector database

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2755977/), step Size check on d12fe87 wanted to post the following:

Warning: This PR changes repo size by 13.50 MiB.

MengnanLi91 committed:
Change to local host machine

hugary1995 commented on .github/workflows/discussionbot-workflow.yml:
```diff
@@ -0,0 +1,35 @@
+    types:
+    # Use locked as trigger event until this chatbot is ready
+      - created
+    #  - locked
```
```suggestion
```

hugary1995 replied:
I wonder if there are other events that should trigger this. Is there an event associated with editing the content?

hugary1995 commented on .github/workflows/discussionbot-workflow.yml:
```diff
@@ -0,0 +1,35 @@
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
```
This is at v4 now.

hugary1995 commented on .github/workflows/discussionbot-workflow.yml:
```diff
@@ -0,0 +1,35 @@
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.10"
```
Is hard-coding python version here okay?

hugary1995 commented on .github/workflows/discussionbot-workflow.yml:
```diff
@@ -0,0 +1,35 @@
+      - name: Install dependencies
+        run: pip install -r python/MooseDiscussion/requirements.txt
+
+      - name: Run discuss bot
```
```suggestion
      - name: Run discussion bot
```

hugary1995 commented on .github/workflows/discussionbot-workflow.yml:
```diff
@@ -0,0 +1,35 @@
+        run: python python/MooseDiscussion/GitHubBot.py
+
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Do you need to explicitly pass this secret? From what I recall github actions use this token by default.

MengnanLi91 replied:
I don't think I need to. I have modified this in the new version

hugary1995 commented on python/MooseDiscussion/GitHubAPI.py:
```diff
@@ -0,0 +1,119 @@
+import argparse
```
We may want a shebang at the top.

hugary1995 commented on python/MooseDiscussion/GitHubBot.py:
```diff
@@ -0,0 +1,187 @@
+from pathlib import Path
```
Consider adding a shebang.

hugary1995 commented on python/MooseDiscussion/IndexGenerator.py:
```diff
@@ -0,0 +1,145 @@
+import json
```
Consider adding a shebang.

hugary1995 reviewed (changes requested):
I share similar concerns about database storage. I have some other minor comments.

hugary1995 commented:
> We are also looking for a fitting name for the discussion bot. If you have any suggestions, please share them here!

Not sure if we should try to adhere to our animal naming convention here. But maybe mechamoose?

![image](https://github.com/user-attachments/assets/31bece08-bc26-4f3f-9095-110a7e8055fa)

MengnanLi91 committed:
try ubuntu-latest instead of linux

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2964184/), step Size check on 6bf6ffb wanted to post the following:

Warning: This PR changes repo size by 13.50 MiB.

MengnanLi91 committed:
Make output looks better

MengnanLi91 committed:
Remove 'Title:'

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2964465/), step Size check on f5dc8b4 wanted to post the following:

Warning: This PR changes repo size by 13.50 MiB.
